bootstrap.lua

```diff
@@ -21,47 +21,134 @@ require 'common/globals'
+            return { loaded = false, data={missing = {missingText}, errored = erroredCASCs} }
+        end
+
+        local cascs = Split(abyss.getConfig("OpenDiablo2", "CASCs"), ",")
+        for _, _casc in ipairs(cascs) do
+            abyss.log("info", string.format("Loading CASC %s...", baseDataDir .. "/"  .. _casc))
+            local loaded = pcall(function()
+                abyss.addLoaderProvider("casc", baseDataDir .. "/"  .. _casc)
+            end)
+            if not loaded then
+                abyss.log("error", string.format("Possible corrupt or missing CASC: %s", baseDataDir .. "/" .. _casc))
+                table.insert(erroredCASCs, baseDataDir .. "/" .. _casc)
+                errored = true
+            end
+        end
+        return {loaded = not errored, data={missing = missingCASCs, errored = erroredCASCs}}
+    end
+
+    local function loadBaseMPQs()
+        local errored = false
+        local toLoadMPQs = {}
+        local missingMPQs = {}
+        local erroredMPQs = {}
+
+        if abyss.fileExists(".disableMPQ") then
+            local missingText = "MPQ LOADING DISABLED BY FILE: .disableMPQ"
+            return { loaded = false, data={missing = {missingText}, errored = erroredMPQs} }
+        end
+
+        local mpqFolderNames = Split(abyss.getConfig("OpenDiablo2", "MPQs"), ",")
+
+        for i in pairs(mpqFolderNames) do
+            local mpqPath = baseDataDir .. "/" .. mpqFolderNames[i]
+            local mpqLoadOrderFileName = mpqPath .. "/" .. ".loadorder"
+            local file = io.open (mpqLoadOrderFileName, "r")
+            local fileData = file:read("a")
+            file:close()
+            local mpqFiles = Split(fileData, ",")
+            for _, mpqFile in ipairs(mpqFiles) do
+                if file_exists(mpqPath .. "/" .. mpqFile) then
+                    table.insert(toLoadMPQs,mpqPath .. "/" .. mpqFile)
+                else
+                    table.insert(missingMPQs, mpqPath .. "/" .. mpqFile)
+                    errored = true
+                end
+            end
+        end
+
+        -- If no files are missing then really load them
+        for _, _mpq in ipairs(toLoadMPQs) do
+            local mpqPath = baseDataDir .. "/" .. _mpq
+            local loadStr = string.format("Loading MPQ %s...", _mpq)
+            abyss.log("info", loadStr)
+            local loaded = pcall(function() abyss.addLoaderProvider("mpq", _mpq) end)
+            if not loaded then
+                table.insert(erroredMPQs, _mpq)
+                abyss.log("error", string.format("Possible corrupt MPQ: %s", _mpq))
+                errored = true
+            end
+        end
+        local result = {loaded = not errored, data={ missing = missingMPQs, errored = erroredMPQs}}
+        return result
+    end
 
-abyss.addLoaderProvider("filesystem", "./fallback-data")
+    abyss.addLoaderProvider("filesystem", "./DATA/FS-fallback")
```

Comment:
you broke d2r mode

the fallback should be after casc, because we prefer json layout files from casc over the fallback ones - fallback ones are not complete